File: Behavioral/Strategy/Kotlin/Draw.kt
// soyut strategy
interface Brush
{
    fun Paint()
}

// Somut strategy siniflari
class SolidBrush: Brush
{
    override fun Paint() {
        print("Tek renkli boyama")
    }
}

class LinearGradientBrush: Brush
{
    override fun Paint() {
        print("cok renkli cizgisel gecisli boyama")
    }
}

class RadialGradientBrush: Brush
{
    override fun Paint() {
        println("cok renkli yuvarlak gecisli boyama")
    }
}

// Context
class Rectangle(brush: Brush){
    private var brush: Brush

    init {
        this.brush = brush
    }

    fun Draw() {
        brush.Paint()
        print("Dikdortgen")
    }
}

// test
object Program
{
    @JvmStatic fun main(args:Array<String>)
    {
        var b1 = SolidBrush()
        var b2 = LinearGradientBrush()
        var b3 = RadialGradientBrush()

        var r = Rectangle(b3)
        r.Draw()

    }
}

File: Behavioral/Strategy/Swift/Draw.swift
import Foundation

// soyut strategy
protocol Brush
{
    func Paint()
}

// somut strategy siniflari
class SolidBrush: Brush
{
    func Paint() {
        print("tek renkli boyama")
    }
}

class LinearGradientBrush: Brush
{
    func Paint() {
        print("cok renkli cizgisel gecisli boyama")
    }
}

class RadialGradientBrush: Brush
{
    func Paint() {
        print("cok renkli yuvarlak gecisli boyama")
    }
}


// Context

class Rectangle
{
    private var brush:Brush

    init(brush:Brush) {
        self.brush = brush
    }

    func Draw() {
        brush.Paint()
        print("Dikdortgen")
    }

}


////// test /////////
var b1 = SolidBrush()
var b2 = LinearGradientBrush()
var b3 = RadialGradientBrush()

var r = Rectangle(brush: b3)
r.Draw()
